### PR TITLE
Initial Linux/LoongArch64 Port

### PR DESCRIPTION
**Pull request by xiangzhai**
#### 647e67b23883960fef8890465c0f70d7ab6e63f1
<pre>
Initial Linux/LoongArch64 Port
<a href="https://bugs.webkit.org/show_bug.cgi?id=244664">https://bugs.webkit.org/show_bug.cgi?id=244664</a>

Reviewed by Michael Catanzaro.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/WTF/wtf/PageBlock.h:
* Source/WTF/wtf/PlatformCPU.h:
* Source/WTF/wtf/dtoa/utils.h:
* Source/cmake/WebKitCommon.cmake:

Canonical link: <a href="https://commits.webkit.org/254155@main">https://commits.webkit.org/254155@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/2db21334f8a39daab4cd9b7244e16be53931434f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88218 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/69/builds/32540 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18996 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97414 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/152880 "Built successfully and passed tests") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/64/builds/30930 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26744 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80380 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92081 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93827 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/64/builds/30930 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74977 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24742 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/64/builds/30930 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79812 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/67715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79967 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/28585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13772 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/73721 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/28636 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14758 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26168 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2914 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/31753 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37670 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/76565 "Built successfully") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/30704 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33926 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16991 "Passed tests") |

### CODE REVIEW

File: Source/JavaScriptCore/heap/HandleBlock.h [hunk 1]
```diff
@@ -39,7 +39,11 @@ class HandleBlock : public DoublyLinkedListNode<HandleBlock> {
     static void destroy(HandleBlock*);
     static HandleBlock* blockFor(HandleNode*);
 
+#if defined(__mips__) || defined(__loongarch__)
+    static constexpr size_t blockSize = 16 * KB;
+#else
     static constexpr size_t blockSize = 4 * KB;
+#endif
```
Comment (mcatanzaro) [inline Source/JavaScriptCore/heap/HandleBlock.h]: Can you explain why this is needed, please? We already have MIPS support and it doesn't seem to need this. There's also no apparent relation to anything processor-specific?

File: Source/JavaScriptCore/heap/WeakBlock.h [hunk 2]
```diff
@@ -41,8 +41,11 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(WeakBlock);
 class WeakBlock : public DoublyLinkedListNode<WeakBlock> {
 public:
     friend class WTF::DoublyLinkedListNode<WeakBlock>;
+#if defined(__mips__) || defined(__loongarch__)
+    static constexpr size_t blockSize = 1024; // 1/16 of MarkedBlock size
+#else
     static constexpr size_t blockSize = 256; // 1/16 of MarkedBlock size
-
+#endif
```
Comment (mcatanzaro) [inline Source/JavaScriptCore/heap/WeakBlock.h]: Same question here. (I see you found a stale comment, as 256 is definitely not guaranteed to be 1/16 of the MarkedBlock size. We should fix that, but the fix probably shouldn't be processor-specific like this.)

File: Source/cmake/WebKitCommon.cmake [hunk 3]
```diff
@@ -67,6 +67,8 @@ if (NOT HAS_RUN_WEBKIT_COMMON)
         set(COMPILER_IS_CLANG ON)
     endif ()
 
+    option(USE_LIBCXX "Use llvm libc++" OFF)
```
Comment (mcatanzaro) [inline Source/cmake/WebKitCommon.cmake]: Mmm, I kinda suspect that most people who want libc++ will either pass the -stdlib=libc++ flag manually, or will get it by default, but I suppose it's harmless. It's surely not needed for porting to loongarch64, though, so I would split this out into a separate commit.

File: Source/cmake/WebKitCompilerFlags.cmake [hunk 4]
```diff
@@ -196,6 +196,10 @@ if (LTO_MODE AND COMPILER_IS_CLANG)
     set(CMAKE_MODULE_LINKER_FLAGS "-flto=${LTO_MODE} ${CMAKE_MODULE_LINKER_FLAGS}")
 endif ()
 
+if (COMPILER_IS_CLANG AND USE_LIBCXX)
+    set(CMAKE_CXX_FLAGS "-std=c++20 -stdlib=libc++ ${CMAKE_CXX_FLAGS}")
```
Comment (mcatanzaro) [inline Source/cmake/WebKitCompilerFlags.cmake]: You shouldn't mess with `-std=` here, though. That's not related to the chosen libc, and it's redundant with wherever else it gets set. Setting it here is asking for trouble next time we bump to a newer C++ version and forget to update it here.

**Review by mcatanzaro — CHANGES_REQUESTED** [review]
This looks mostly uncontroversial, although I've requested a few changes. Except for one thing: you have some big unexplained changes in ArgumentCoders.h. Can you please explain in detail what you're doing there? It looks more likely to be related to trouble with a particular compiler version rather than CPU architecture?

File: Source/cmake/WebKitCommon.cmake (hunk 3 above)
Comment (mcatanzaro) [reply Source/cmake/WebKitCommon.cmake]: Well actually, I see this is only needed if the platform default stdlib is libstdc++. So that means this option ensures that the compiled WebKit is incompatible with other system libraries. It's probably better not to provide this, right? It's an unusual thing to want to do, and you can always pass the flag manually if you really want it.

Comment (mcatanzaro) [reply Source/cmake/WebKitCommon.cmake]: Ah, I see you split this out into https://github.com/WebKit/WebKit/pull/3904.

**Comment by xiangzhai** [comment]
Hi  @mcatanzaro 

Thanks for your kind response!

I will update my patch.

Thanks,
Leslie Zhai

File: Source/JavaScriptCore/heap/WeakBlock.h (hunk 2 above)
Comment (xiangzhai) [reply Source/JavaScriptCore/heap/WeakBlock.h]: > We should fix that

I am familiar with OpenJDK hotspot JIT compilers and [LLVM](https://github.com/llvm/llvm-project/commits?author=xiangzhai).  But I am an apprentice in WebKit world :)  I will fix the "PAGE SIZE" issue,  due to LoongArch64 default page size is 16K, much more platform independent.

**Comment by xiangzhai** [comment]
Hi  @mcatanzaro  

> Except for one thing: you have some big unexplained changes in ArgumentCoders.h

When use llvm's libc++:

```
set(CMAKE_CXX_FLAGS "-std=c++20 -stdlib=libc++ ${CMAKE_CXX_FLAGS}")
```

libcxx 13 thrown tuple failed to conversion related error:

```
FAILED: Source/WebKit/CMakeFiles/WebKit.dir/__/__/DerivedSources/WebKit/WebInspectorUIProxyMessageReceiver.cpp.o 
/usr/bin/c++ -DBUILDING_GTK__=1 -DBUILDING_WEBKIT -DBUILDING_WITH_CMAKE=1 -DBUILDING_WebKit -DBWRAP_EXECUTABLE=\"/usr/bin/bwrap\" -DDATADIR=\"/home/loongson/webkit-loongarch64-build/share\" -DDBUS_PROXY_EXECUTABLE=\"/usr/bin/xdg-dbus-proxy\" -DGETTEXT_PACKAGE=\"WebKit2GTK-4.0\" -DHAVE_CONFIG_H=1 -DJSC_GLIB_API_ENABLED -DLIBDIR=\"/home/loongson/webkit-loongarch64-build/lib\" -DLOCALEDIR=\"/home/loongson/webkit-loongarch64-build/share/locale\" -DPAS_BMALLOC=1 -DPKGLIBEXECDIR=\"/home/loongson/webkit-loongarch64-build/libexec/webkit2gtk-4.0\" -DSTATICALLY_LINKED_WITH_PAL -DSTATICALLY_LINKED_WITH_WebCore -DWEBKIT2_COMPILATION -DWEBKITGTK_API_VERSION_STRING=\"4.0\" -DWEBKIT_DOM_USE_UNSTABLE_API -DWebKit_EXPORTS -I/home/loongson/private/WebKit/build/WebKit2Gtk/Headers ... -isystem /usr/include/enchant -isystem /usr/include/gio-unix-2.0 -isystem /usr/include/libmount -isystem /usr/include/blkid -isystem /usr/include/uuid -isystem /usr/include/glib-2.0 -isystem /usr/lib/loongarch64-linux-gnu/glib-2.0/include -isystem /usr/include/gstreamer-1.0 -isystem /usr/include/gtk-3.0 -isystem /usr/include/at-spi2-atk/2.0 -isystem /usr/include/at-spi-2.0 -isystem /usr/include/dbus-1.0 -isystem /usr/lib/loongarch64-linux-gnu/dbus-1.0/include -isystem /usr/include/cairo -isystem /usr/include/libdrm -isystem /usr/include/pango-1.0 -isystem /usr/include/harfbuzz -isystem /usr/include/fribidi -isystem /usr/include/atk-1.0 -isystem /usr/include/pixman-1 -isystem /usr/include/freetype2 -isystem /usr/include/libpng16 -isystem /usr/include/gdk-pixbuf-2.0 -isystem /usr/include/libsoup-2.4 -isystem /usr/include/gtk-3.0/unix-print -isystem /usr/include/libxml2 -isystem /usr/include/openjpeg-2.3 -std=c++20 -stdlib=libc++ -fdiagnostics-color=always -fcolor-diagnostics -Wextra -Wall -pipe -Wno-noexcept-type -Wno-psabi -Wno-misleading-indentation -Wno-parentheses-equality -Qunused-arguments -Wwrite-strings -Wundef -Wpointer-arith -Wmissing-format-attribute -Wformat-security -Wcast-align -Wno-tautological-compare  -fno-strict-aliasing -fno-exceptions -fno-rtti -O2 -g -DNDEBUG -fPIC -fvisibility=hidden  -include /home/loongson/private/WebKit/Source/WebKit/WebKit2Prefix.h -Wno-unused-parameter -pthread -MD -MT Source/WebKit/CMakeFiles/WebKit.dir/__/__/DerivedSources/WebKit/WebInspectorUIProxyMessageReceiver.cpp.o -MF Source/WebKit/CMakeFiles/WebKit.dir/__/__/DerivedSources/WebKit/WebInspectorUIProxyMessageReceiver.cpp.o.d -o Source/WebKit/CMakeFiles/WebKit.dir/__/__/DerivedSources/WebKit/WebInspectorUIProxyMessageReceiver.cpp.o -c /home/loongson/private/WebKit/build/DerivedSources/WebKit/WebInspectorUIProxyMessageReceiver.cpp
In file included from <built-in>:1:
In file included from /home/loongson/private/WebKit/Source/WebKit/WebKit2Prefix.h:72:
In file included from /usr/lib/llvm-13/bin/../include/c++/v1/algorithm:653:
In file included from /usr/lib/llvm-13/bin/../include/c++/v1/functional:495:
In file included from /usr/lib/llvm-13/bin/../include/c++/v1/__functional/bind_front.h:14:
In file included from /usr/lib/llvm-13/bin/../include/c++/v1/__functional/perfect_forward.h:14:
/usr/lib/llvm-13/bin/../include/c++/v1/tuple:1505:12: error: no viable conversion from returned value of type 'tuple<IPC::Attachment &&>' to function return type 'tuple<IPC::Attachment>'
    return __tuple_cat<tuple<>, __tuple_indices<>,
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/loongson/private/WebKit/Source/WebKit/Platform/IPC/ArgumentCoders.h:534:21: note: in instantiation of function template specialization 'std::tuple_cat<std::tuple<IPC::Attachment>, std::tuple<>>' requested here
        return std::tuple_cat(std::make_tuple(WTFMove(*optional)), WTFMove(*remainder));
                    ^
/home/loongson/private/WebKit/Source/WebKit/Platform/IPC/ArgumentCoders.h:557:43: note: in instantiation of function template specialization 'IPC::TupleDecoder<IPC::Attachment>::decode<IPC::Decoder>' requested here
        return TupleDecoder<Elements...>::decode(decoder);
                                          ^
/home/loongson/private/WebKit/Source/WebKit/Platform/IPC/Decoder.h:126:40: note: in instantiation of function template specialization 'IPC::ArgumentCoder<std::tuple<IPC::Attachment>>::decode<IPC::Decoder>' requested here
            std::optional<T> t { Impl::decode(*this) };
                                       ^
/home/loongson/private/WebKit/Source/WebKit/Platform/IPC/HandleMessage.h:254:30: note: in instantiation of function template specialization 'IPC::Decoder::decode<std::tuple<IPC::Attachment>>' requested here
    auto arguments = decoder.decode<typename CodingType<typename T::Arguments>::Type>();
                             ^
/home/loongson/private/WebKit/build/DerivedSources/WebKit/WebInspectorUIProxyMessageReceiver.cpp:98:21: note: in instantiation of function template specialization 'IPC::handleMessage<Messages::WebInspectorUIProxy::SetFrontendConnection, WebKit::WebInspectorUIProxy, void (WebKit::WebInspectorUIProxy::*)(IPC::Attachment)>' requested here
        return IPC::handleMessage<Messages::WebInspectorUIProxy::SetFrontendConnection>(connection, decoder, this, &WebInspectorUIProxy::setFrontendConnection);
                    ^
/usr/lib/llvm-13/bin/../include/c++/v1/tuple:663:5: note: candidate constructor not viable: no known conversion from 'typename __tuple_cat_return_ref<tuple<> &&, tuple<Attachment> &&, tuple<> &&>::type' (aka 'tuple<IPC::Attachment &&>') to 'const std::tuple<IPC::Attachment> &' for 1st argument
    tuple(const tuple&) = default;
    ^
/usr/lib/llvm-13/bin/../include/c++/v1/tuple:664:5: note: candidate constructor not viable: no known conversion from 'typename __tuple_cat_return_ref<tuple<> &&, tuple<Attachment> &&, tuple<> &&>::type' (aka 'tuple<IPC::Attachment &&>') to 'std::tuple<IPC::Attachment> &&' for 1st argument
    tuple(tuple&&) = default;
    ^
/usr/lib/llvm-13/bin/../include/c++/v1/tuple:527:5: note: candidate template ignored: substitution failure [with _And = _And]: no member named '_EnableIfImpl' in 'std::_MetaBase<false>'
    tuple(const _Tp& ... __t)
    ^
/usr/lib/llvm-13/bin/../include/c++/v1/tuple:606:5: note: candidate template ignored: substitution failure [with _Up = <std::tuple<IPC::Attachment &&>>]: no member named '_EnableIfImpl' in 'std::_MetaBase<false>'
    tuple(_Up&&... __u)
    ^
/usr/lib/llvm-13/bin/../include/c++/v1/tuple:704:5: note: candidate template ignored: substitution failure [with _Up = <IPC::Attachment &&>]: no member named '_EnableIfImpl' in 'std::_MetaBase<false>'
    tuple(const tuple<_Up...>& __t)
    ^
/usr/lib/llvm-13/bin/../include/c++/v1/tuple:770:5: note: candidate template ignored: substitution failure [with _Up = <IPC::Attachment &&>]: no member named '_EnableIfImpl' in 'std::_MetaBase<false>'
    tuple(tuple<_Up...>&& __t)
    ^
/usr/lib/llvm-13/bin/../include/c++/v1/tuple:836:5: note: candidate template ignored: could not match 'pair' against 'tuple'
    tuple(const pair<_Up1, _Up2>& __p)
    ^
/usr/lib/llvm-13/bin/../include/c++/v1/tuple:905:5: note: candidate template ignored: could not match 'pair' against 'tuple'
    tuple(pair<_Up1, _Up2>&& __p)
    ^
/usr/lib/llvm-13/bin/../include/c++/v1/tuple:544:14: note: explicit constructor is not a candidate
    explicit tuple(const _Tp& ... __t)
             ^
/usr/lib/llvm-13/bin/../include/c++/v1/tuple:622:14: note: explicit constructor is not a candidate
    explicit tuple(_Up&&... __u)
             ^
/usr/lib/llvm-13/bin/../include/c++/v1/tuple:717:14: note: explicit constructor is not a candidate
    explicit tuple(const tuple<_Up...>& __t)
             ^
/usr/lib/llvm-13/bin/../include/c++/v1/tuple:783:14: note: explicit constructor is not a candidate
    explicit tuple(tuple<_Up...>&& __t)
             ^
/usr/lib/llvm-13/bin/../include/c++/v1/tuple:851:14: note: explicit constructor is not a candidate
    explicit tuple(const pair<_Up1, _Up2>& __p)
             ^
/usr/lib/llvm-13/bin/../include/c++/v1/tuple:920:14: note: explicit constructor is not a candidate
    explicit tuple(pair<_Up1, _Up2>&& __p)
             ^
1 error generated.
```

So I just reverted the commit `34daeac7dc0fb6f1d9421c5c6c8cdf92cdc0abc5`. But I reverted my revert now :)  Please review it again.

BTW: I rebased to WebKit-7613.3.9.3.3 for another go, owing to it is able to use GCC 8.3 and libstdc++,  perhaps GNU toolchain will not reproduce the tuple conversion issue.

Thanks,
Leslie Zhai

**Comment by xiangzhai** [comment]
Hi  @mcatanzaro 

> BTW: I rebased to WebKit-7613.3.9.3.3 for another go, owing to it is able to use GCC 8.3 and libstdc++, perhaps GNU toolchain will not reproduce the tuple conversion issue.

Good news not reproduced:   https://github.com/xiangzhai/WebKit/commit/b6f682c68283ba606d5a039db7881ad7540098d9  Because there is no such commit `Simplify IPC encoding, decoding of tuples` in the tag  WebKit-7613.3.9.3.3.

Cheers,
Leslie Zhai

**Comment by mcatanzaro** [comment]
I guess we will need a JavaScriptCore developer to review the changes in HandleBlock.h and WeakBlock.h. I'm not sure why loongarch needs to be different from all other architectures if (a) what we have now works on architectures as diverse as ppc64le and s390x, and (b) loongarch uses the same 16 KiB MarkedBlock size that Apple does. Perhaps @Constellation or @MenloDorian? Notably, the preexisting comment in WeakBlock.h is not correct. It says WeakBlock's size of 256 bytes is 1/16 of MarkedBlock size, but MarkedBlock size is minimum 16 KiB and max 64 KiB, depending on architecture. Looks like this comment was written when MarkedBlock was only 4 KiB.

All the other changes in this pull request are now clearly fine. If you prefer, we could land these now, and you could follow up on HandleBlock.h and WeakBlock.h separately. Note the pull request must contain only one commit to land, so you'll need to rewrite the history with `git rebase -i` and then force push with `git push -f`.

**Comment by xiangzhai** [comment]
Hi  @mcatanzaro  

> All the other changes in this pull request are now clearly fine. If you prefer, we could land these now, and you could follow up on HandleBlock.h and WeakBlock.h separately

Done.


> Note the pull request must contain only one commit to land, so you'll need to rewrite the history with `git rebase -i` and then force push with `git push -f`

```
git rebase -i loongarch-port
git push -f origin loongarch-port
```

Please point out my fault.

Thanks,
Leslie Zhai

**Comment by xiangzhai** [comment]
> Note the pull request must contain only one commit to land

Done.

Thank  @theaoqi  pointed out my fault :wine_glass: 

Cheers,
Leslie Zhai

**Review by mcatanzaro — APPROVED** [review]

**Comment by mcatanzaro** [comment]
Please change the commit message to look exactly like this:

```
Initial Linux/LoongArch64 Port
https://bugs.webkit.org/show_bug.cgi?id=244664

Reviewed by NOBODY (OOPS!).

* Source/JavaScriptCore/CMakeLists.txt:
* Source/WTF/wtf/PageBlock.h:
* Source/WTF/wtf/PlatformCPU.h:
* Source/WTF/wtf/dtoa/utils.h:
* Source/cmake/WebKitCommon.cmake:
```

(I tried to do this for you, but you didn't allow commit access to your fork. Probably best to do it yourself anyway.)

**Comment by xiangzhai** [comment]
Hi  @mcatanzaro 

> Please change the commit message to look exactly like this:

Done.

Cheers,
Leslie Zhai

**Review by mcatanzaro — APPROVED** [review]

**Comment by aperezdc** [comment]
The libc++ issue is now in Bugzilla as [issue #245044](https://bugs.webkit.org/show_bug.cgi?id=245044)

**Comment by xiangzhai** [comment]
> The libc++ issue is now in Bugzilla as [issue #245044](https://bugs.webkit.org/show_bug.cgi?id=245044)

Thanks a lot!

And at present we just revert the commit `34daeac7dc0fb6f1d9421c5c6c8cdf92cdc0abc5` for LoongArch64:  https://github.com/loongson/WebKit/commit/5657170d7bd1492a522d0c35269d09b90ef78aa2#diff-92c56ce66bb066fa2340e34258930e3f33b5742330cfc3e460426d0df25cee3a
